pimd/pim_nb_config.c

```diff
@@ -125,12 +122,16 @@ static void pim_if_membership_refresh(struct interface *ifp)
 		} /* scan group sources */
 	}	 /* scan igmp groups */
 #else
-	sg_start = gm_sgs_first(gm_ifp->sgs);
+	gm_ifp = pim_ifp->mld;
 
-	frr_each_from (gm_sgs, gm_ifp->sgs, sg, sg_start) {
-		if (!in6_multicast_nofwd(&sg->sgaddr.grp)) {
-			pim_ifchannel_local_membership_add(
-				ifp, &sg->sgaddr, false /*is_vxlan*/);
+	if (gm_ifp) {
```

Comment:
There is some other problem in the code due to which this has become NULL.  Can you please check and fix that.

If you see line 87 above, it makes sure if MLD is not enabled it should be returned. So why that is true and MLD pointer is NULL. We should not fix like this.

Reply:
Hi Mobash,
Thank you for the comment.
This crash happens when we apply MLD/PIMV6 configuration on an interface which is not configured/present.
In this case, currently we create pim_ifp structure with pim_enable and gm_enable = true. But we won't create mld structure. 
So NULL check will help here to prevent the crash.
Here, I am allowing the other things similar to IGMP.
Regards,
Sarita

Reply:
The below code already present in Line 87 should ensure MLD is not enabled. Adding NULL check would hide the actual problem.
if (!pim_ifp->gm_enable)
		return;

Reply:
As I mentioned earlier, we enable this flag gm_enable for MLD similar to IGMP.
But we don't create MLD structure similar.
If we dont enable the enable the flag gm_enable for MLD, it will be different behaviour than IGMP.
Are you suggesting we should not enable gm_enable flag MLD ?

Reply:
No I am not asking you to do that. What I am saying is the actual problem is somewhere else, let's try to find out and fix it. pim_if_membership_refresh function is called for refreshing the membership and not for enabling mld. Prior to this function call, the mld pointer should be filled. Generally it is done in gm_ifp_update function. So please check how mld pointer has become NULL when we are at this function.

Reply:
The mld core is happening when the interface is not configured in linux and user try to configure ipv6 mld and ipv6 pim command. 

If an interface is not configured in linux, the interface will have if index = 0 and mroute_vif_index = -1.
Currently if we enable PIM/MLD on this interface, we create pim_ifp with pim_enable and gm_enable = True.
But we don't create MLD structure gm_ifp for this interface. 
Please find the code for this in gm_ifp_update().
        if (!if_is_operative(ifp) || !pim_ifp->pim || 
            pim_ifp->mroute_vif_index < 0) {
                gm_ifp_teardown(ifp);
                return;
        }

pim_if_membership_refresh() API should check if MLD structure gm_ifp is present or not along with gm_enable and pim_enabled flag.

Reply:
I have verified with the fix.

- Interface ens256 is not configured in linux. 
- Enable ipv6 mld. 
- MLD structure is not created for this interface. Verified using "show ipv6 mld interface " command.

frr(config-if)# do show ipv6 mld interface 
 Interface  State  Address                   V  Querier  QuerierIp                 Query Timer   Uptime        

- Configure interface ens256 in linux.
- MLD structure is created for this interface. Verified using "show ipv6 mld interface " command.

frr# show ipv6 mld interface 
 Interface  State  Address                   V  Querier  QuerierIp                 Query Timer   Uptime        
 ens256     up     fe80::250:56ff:feb7:e3a1  2  local    fe80::250:56ff:feb7:e3a1  00:00:59.082  00:10:30.161

Reply:
Please move this mld null pointer check to line 87 where gm_enable flag is checked so that even in future if someone adds any code in this function this check again is not required.